File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        if (s.length() == 0) return 0;
        vector<int> dp(256, 0);
        int maxi = 1;
        int f = 0;
        for(int i=0; i<s.length(); i++){
            dp[s[i]]++;
            while(dp[s[i]] > 1){
                dp[s[f]]--;
                f++;
            }
            maxi = max(maxi, i-f+1);
        }
        return maxi;
    }
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        charset = set()
        l = r = 0
        length = 0
        for r in range(len(s)):
            if s[r] in charset:
                while l < r and s[l] != s[r]:
                    charset.remove(s[l])
                    l += 1
                charset.remove(s[r])
                l += 1
            charset.add(s[r])
            length = max(length, len(charset))

        return length
            

        